nav2_route/include/nav2_route/utils.hpp

```diff
@@ -59,102 +59,89 @@ inline visualization_msgs::msg::MarkerArray toMsg(
-    curr_marker.pose.position.x = graph[i].coords.x;
-    curr_marker.pose.position.y = graph[i].coords.y;
-    curr_marker.scale = getSphereSize();
-    curr_marker.color = getSphereColor();
-    msg.markers.push_back(curr_marker);
-
-    // Add text
-    curr_marker.ns = "route_graph_ids";
-    curr_marker.id = marker_idx++;
-    curr_marker.type = visualization_msgs::msg::Marker::TEXT_VIEW_FACING;
-    curr_marker.pose.position.x = graph[i].coords.x + 0.07;
-    curr_marker.pose.position.y = graph[i].coords.y;
-    curr_marker.text = std::to_string(graph[i].nodeid);
-    curr_marker.scale.z = 0.1;
-    msg.markers.push_back(curr_marker);
-
-    for (unsigned int j = 0; j != graph[i].neighbors.size(); j++) {
-      curr_marker.ns = "route_graph";
-      curr_marker.id = marker_idx++;
-      curr_marker.type = visualization_msgs::msg::Marker::LINE_LIST;
-      curr_marker.pose.position.x = 0;  // Set to 0 since points are relative to this frame
-      curr_marker.pose.position.y = 0;  // Set to 0 since points are relative to this frame
-      curr_marker.points.resize(2);
-      curr_marker.points[0].x = graph[i].coords.x;
-      curr_marker.points[0].y = graph[i].coords.y;
-      curr_marker.points[1].x = graph[i].neighbors[j].end->coords.x;
-      curr_marker.points[1].y = graph[i].neighbors[j].end->coords.y;
-      curr_marker.scale.x = 0.03;
-      curr_marker.color = getLineColor();
-      msg.markers.push_back(curr_marker);
-      curr_marker.points.clear();  // Reset for next node marker
-
-      // Add text
-      curr_marker.ns = "route_graph_ids";
-      curr_marker.id = marker_idx++;
-      curr_marker.type = visualization_msgs::msg::Marker::TEXT_VIEW_FACING;
-      curr_marker.pose.position.x =
-        graph[i].coords.x + ((graph[i].neighbors[j].end->coords.x - graph[i].coords.x) / 2.0) +
-        0.07;
+  visualization_msgs::msg::Marker nodes_marker;
+  nodes_marker.header.frame_id = frame;
+  nodes_marker.header.stamp = now;
+  nodes_marker.action = 0;
+  nodes_marker.ns = "route_graph_nodes";
+  nodes_marker.type = visualization_msgs::msg::Marker::POINTS;
+  nodes_marker.scale.x = 0.5;
+  nodes_marker.scale.y = 0.5;
+  nodes_marker.scale.z = 0.5;
+  nodes_marker.color.r = 1.0;
+  nodes_marker.color.a = 1.0;
+  nodes_marker.points.reserve(graph.size());
+
+  visualization_msgs::msg::Marker edges_marker;
+  edges_marker.header.frame_id = frame;
+  edges_marker.header.stamp = now;
+  edges_marker.action = 0;
+  edges_marker.ns = "route_graph_edges";
+  edges_marker.type = visualization_msgs::msg::Marker::LINE_LIST;
+  edges_marker.scale.x = 0.1;  // Line width
+  edges_marker.color.g = 1.0;
+  edges_marker.color.a = 0.5;  // Semi-transparent green so bidirectional connections stand out
+  // Each edge has 2 points, each node is likely to have at least 2 edges.
+  // This likely under-reserves but saves some initial reallocations
+  edges_marker.points.reserve(graph.size() * 2 * 2);
```

Comment:
```suggestion
  edges_marker.points.reserve(graph.size() * 4);
```

Also known as `4` :wink: